pennylane/math/matrix_manipulation.py

```diff
@@ -104,44 +106,99 @@ def expand_matrix(base_matrix, wires, wire_order=None, sparse_format="csr"):
+    wires = list(wires)
+    wire_indices = [wire_order.index(wire) for wire in wires]
+    subset_wire_order = wire_order[min(wire_indices) : max(wire_indices) + 1]
+    wire_difference = list(set(subset_wire_order) - set(wires))
+    expanded_wires = wires + wire_difference
+
+    # expand matrix if needed
+    if wire_difference:
+        if batch_dim is not None:
+            batch_matrices = [
+                qml.math.kron(batch, eye_interface(len(wire_difference)), like=interface)
+                for batch in base_matrix
+            ]
+            base_matrix = qml.math.stack(batch_matrices, like=interface)
+        else:
+            base_matrix = qml.math.kron(
+                base_matrix, eye_interface(len(wire_difference)), like=interface
+            )
+
+    num_wires = len(subset_wire_order)
+
+    # reshape matrix to match wire values e.g. mat[0, 0, 0, 0] = <0000|mat|0000>
+    # with this reshape we can easily swap wires
+    shape = [batch_dim] + [2] * (num_wires * 2) if batch_dim else [2] * (num_wires * 2)
+    mat = qml.math.reshape(base_matrix, shape)
+
+    # compute the permutations needed to match wire order
+    perm = [expanded_wires.index(wire) for wire in subset_wire_order]
+    perm += [p + num_wires for p in perm]
     if batch_dim:
-        perm = [p + 1 for p in perm]
-        sources = [s + 1 for s in sources]
+        perm = [0] + [p + 1 for p in perm]
 
-    mat = qml.math.moveaxis(mat_tensordot, sources, perm)
-    shape = [batch_dim] + [2 ** len(wire_order)] * 2 if batch_dim else [2 ** len(wire_order)] * 2
+    if perm != list(range(num_wires * 2)):
+        mat = qml.math.transpose(mat, axes=perm)
+
+    # reshape back
+    shape = [batch_dim] + [2**num_wires] * 2 if batch_dim else [2**num_wires] * 2
     mat = qml.math.reshape(mat, shape)
 
+    # expand the matrix even further by adding the missing wires
+    if len(wire_order) > len(expanded_wires):
+        mats = []  # list of tuples with (Id,) and (base_matrix,) or (mat1, mat2, ...) if batch_dim
+        op_wires_in_list = False
+        identity_count = 0
+        for wire in wire_order:
+            if wire not in expanded_wires:
+                identity_count += 1
+            elif not op_wires_in_list:
+                if identity_count > 0:
+                    mats.append((eye_interface(identity_count),))
+                    identity_count = 0
+                mats.append(tuple(mat) if batch_dim else (mat,))
+                op_wires_in_list = True
+
+        if identity_count > 0:
+            mats.append((eye_interface(identity_count),))
+
+        # itertools.product will create a tuple of matrices for each different batch
+        mats_list = list(itertools.product(*mats))
+        # here we compute the kron product of each different tuple and stack them back together
+        expanded_batch_matrices = [
+            reduce(
+                lambda i, j: qml.math.kron(
+                    i.contiguous() if interface == "torch" else i,
```

Comment:
I know you left a comment in the PR, but maybe there should also be a source code comment about `contiguous`. Imagining someone looking at this in 6 months and trying to figure out why we had to do this.

Reply:
Done!